src/hypothesis/strategies.py

```diff
@@ -1575,4 +1578,204 @@ def deferred(definition):
+            # \\?\UNC\[server]\[sharename]\
+            builds(lambda *x: tp().join(x),
+                   sep, sep, just(tp('?')), sep, just(tp('UNC')), sep, name,
+                   sep, name, sep),
+            # \\.\[physical_device]\
+            builds(lambda *x: tp().join(x),
+                   sep, sep, just(tp('.')), sep, name, sep),
+        ])
+
+        return draw(one_of(relative, drive, extended, network))
+
+
+@defines_strategy
+@composite
+def fspaths(draw, allow_pathlike=None, allow_existing=False):
+    """A strategy which generates filesystem path values.
+
+    The generated values include everything which the builtin
+    :func:`python:open` function accepts i.e. which won't lead to
+    :exc:`ValueError` or :exc:`TypeError` being raised.
+
+    Note that the range of the returned values depends on the operating
+    system, the Python version, and the filesystem encoding as returned by
+    :func:`sys.getfilesystemencoding`.
+
+    :param allow_pathlike:
+        By default makes the strategy also include objects implementing the
+        :class:`python:os.PathLike` interface when Python >= 3.6 is used. If
+        :obj:`python:False` no pathlike objects will be generated. If
+        :obj:`python:True` pathlike will be generated (Python >= 3.6 required)
+
+    :type allow_pathlike: :obj:`python:bool` or :obj:`python:None`
+
+    :param bool allow_existing:
+        If paths which exist on the filesystem should be generated.
+
+    .. versionadded:: 3.15
+
+    """
+    has_pathlike = hasattr(os, 'PathLike')
+
+    if allow_pathlike is None:
+        allow_pathlike = has_pathlike
+    if allow_pathlike and not has_pathlike:  # pragma: no cover
+        raise InvalidArgument(
+            'allow_pathlike: os.PathLike not supported, use None instead '
+            'to enable it only when available')
+
+    result_type = draw(sampled_from([bytes, text_type]))
+
+    def tp(s=''):
+        return _str_to_path(s, result_type)
+
+    special_component = sampled_from([tp(os.curdir), tp(os.pardir)])
+    normal_component = _filename(result_type)
+    path_component = one_of(normal_component, special_component)
+    extension = normal_component.map(lambda f: tp(os.extsep) + f)
+    root = _path_root(result_type)
+
+    def optional(st):
+        return one_of(st, just(result_type()))
+
+    sep = sampled_from([os.sep, os.altsep or os.sep]).map(tp)
+    path_part = builds(lambda s, l: s.join(l), sep, lists(path_component))
+    main_strategy = builds(lambda *x: tp().join(x),
+                           optional(root), path_part, optional(extension))
+
+    if not allow_existing:
+
+        def check_not_exists(path):
+            # os.path.exists returns False for broken symlinks, so use lstat
+            try:
+                os.lstat(path)
+            except OSError:  # pragma: no cover
+                return True
+            except ValueError:  # pragma: no cover
```

Comment:
> Does this mean that we're filtering out too long paths if and only if we have allow_existing enabled?

Yes. Too long paths seemed useful as they can happen when concatenating values in Python. But filtering is fine with me.

I just noticed that not all Python versions throw ValueError, but WindowsError and that the error codes differ between wine and real Windows. That'll require some tests.